uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-infinispan/src/main/java/org/uberfire/ext/metadata/backend/infinispan/provider/InfinispanContext.java

```diff
@@ -0,0 +1,218 @@
+    private static final String SCHEMAS_CACHE = "schemas";
+    private static final String PROTO_EXTENSION = ".proto";
+    private static final String SCHEMA_PROTO = "schema.proto";
+    private static final String ORG_KIE = "org.kie.";
+    private final RemoteCacheManager cacheManager;
+    private final KieProtostreamMarshaller marshaller = new KieProtostreamMarshaller();
+    private final SchemaGenerator schemaGenerator;
+
+    private final InfinispanConfiguration infinispanConfiguration;
+
+    private static InfinispanContext INSTANCE;
+
+    public static InfinispanContext getInstance() {
+        if (INSTANCE == null) {
+            Map<String, String> properties = new HashMap<String, String>() {{
+                put(HOST,
+                    System.getProperty(HOST,
+                                       "127.0.0.1"));
+                put(PORT,
+                    System.getProperty(PORT,
+                                       "11222"));
+                put(USERNAME,
+                    System.getProperty(USERNAME,
+                                       "user"));
+                put(PASSWORD,
+                    System.getProperty(PASSWORD,
+                                       "user"));
+            }};
+            INSTANCE = new InfinispanContext(properties);
+        }
+        return INSTANCE;
+    }
+
+    public static InfinispanContext getInstance(Map<String, String> properties) {
+        if (INSTANCE == null) {
+            INSTANCE = new InfinispanContext(properties);
+        }
+        return INSTANCE;
+    }
+
+    private InfinispanContext(Map<String, String> properties) {
+        this.infinispanConfiguration = new InfinispanConfiguration();
+        schemaGenerator = new SchemaGenerator();
+        ConfigurationBuilder builder = new ConfigurationBuilder();
+        builder.addServer()
+                .host(properties.get(HOST))
+                .port(Integer.parseInt(properties.get(PORT)))
+                .marshaller(new ProtoStreamMarshaller())
+                .marshaller(marshaller);
+
+        cacheManager = new RemoteCacheManager(builder.build());
+
+        if (!this.getIndices().contains(TYPES_CACHE)) {
+            cacheManager.administration().createCache(TYPES_CACHE,
+                                                      this.infinispanConfiguration.getConfiguration(TYPES_CACHE));
+        }
+
+        if (!this.getIndices().contains(SCHEMAS_CACHE)) {
+            cacheManager.administration().createCache(SCHEMAS_CACHE,
+                                                      this.infinispanConfiguration.getConfiguration(SCHEMAS_CACHE));
+        }
+
+        marshaller.registerMarshaller(new KieProtostreamMarshaller.KieMarshallerSupplier<KObject>() {
+            @Override
+            public String extractTypeFromEntity(KObject entity) {
+                return ORG_KIE + AttributesUtil.toProtobufFormat(entity.getClusterId() + "_" + entity.getType().getName());
```

Comment:
Is it ok that ORG_KIE contains dots, which are not escaped?

Reply:
Yes it is ok, because is part of Protobuf package name